packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -650,68 +650,141 @@ export class FluidDataStoreRuntime
+						});
+
+						// If a non-local operation then go and create the object
+						// Otherwise mark it as officially attached.
+						if (local) {
+							assert(
+								this.pendingAttach.delete(id),
+								0x17c /* "Unexpected attach (local) channel OP" */,
+							);
+						} else {
+							assert(!this.contexts.has(id), 0x17d /* "Unexpected attach channel OP" */);
+
+							const summarizerNodeParams = {
+								type: CreateSummarizerNodeSource.FromAttach,
+								sequenceNumber: message.sequenceNumber,
+								snapshot: attachMessage.snapshot,
+							};
+
+							const remoteChannelContext = this.createRemoteChannelContext(
+								attachMessage,
+								summarizerNodeParams,
+							);
+							this.contexts.set(id, remoteChannelContext);
+						}
+						break;
 					}
-					break;
+					default:
 				}
 
-				case DataStoreMessageType.ChannelOp:
-					this.processChannelOp(message, local, localOpMetadata);
-					break;
-				default:
+				this.emit("op", message);
+			} catch (error) {
+				throw DataProcessingError.wrapIfUnrecognized(
+					error,
+					"fluidDataStoreRuntimeFailedToProcessMessage",
+					message,
+				);
 			}
+		}
+	}
 
-			this.emit("op", message);
-		} catch (error) {
-			throw DataProcessingError.wrapIfUnrecognized(
-				error,
-				"fluidDataStoreRuntimeFailedToProcessMessage",
+	/**
+	 * Process data store messages. The messages here are contiguous messages for this data store in a batch.
+	 * @param messagesWithMetadata - A bunch of contiguous messages for this data store and their metadata.
+	 * @param local - true if the messages were originally generated by the client receiving it.
+	 */
+	public processMessages(
+		messagesWithMetadata: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata: unknown;
+		}[],
+		local: boolean,
+	): void {
+		this.verifyNotClosed();
+
+		/** Bunch contiguous messages of the same type and process them together. */
+		let previousMessageType: DataStoreMessageType | undefined;
+		let bunchedOps: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata: unknown;
+		}[] = [];
+
+		messagesWithMetadata.forEach(({ message, localOpMetadata }) => {
+			// If the type of the message changes while processing the batch, process the previous bunch.
+			const maybeProcessBunch = (messageType: DataStoreMessageType) => {
+				if (previousMessageType === undefined || messageType === previousMessageType) {
+					return;
+				}
+				if (bunchedOps.length > 0) {
+					this.processBunchedMessages(bunchedOps, local);
+					bunchedOps = [];
+				}
+			};
+
+			const currentMessageType = message.type as DataStoreMessageType;
+			maybeProcessBunch(currentMessageType);
+			previousMessageType = currentMessageType;
+			bunchedOps.push({
 				message,
-			);
+				localOpMetadata,
+			});
+		});
+
+		// Process the last bunch.
+		if (bunchedOps.length > 0) {
+			this.processBunchedMessages(bunchedOps, local);
 		}
 	}
 
+	public process(
```

Comment:
Probably kept around of back-compat since it only redirects to processMessages. Would be good to mention in the comments somewhere?